README.md

```diff
@@ -1,22 +1,90 @@
+# Honey Badger Byzantine Fault Tolerant (BFT) consensus algorithm
+
 [![Build Status](https://travis-ci.com/poanetwork/hbbft.svg?branch=master)](https://travis-ci.com/poanetwork/hbbft) 
 [![Gitter](https://badges.gitter.im/poanetwork/hbbft.svg)](https://gitter.im/poanetwork/hbbft?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge)
 
-# About
+Welcome to a [Rust](https://www.rust-lang.org/en-US/) library of the Honey Badger Byzantine Fault Tolerant (BFT) consensus algorithm. The research and protocols for this algorithm are explained in detail in "[The Honey Badger of BFT Protocols](https://eprint.iacr.org/2016/199.pdf)" by Miller et al.
```

Comment:
Maybe we should add a list of points in which we deviate from the paper. Currently those would be, I think:

* We use pairing-based cryptography instead of Gap Diffie-Hellman groups. But the principles are the same, and I think even Andrew Miller's Python implementation does that.
* We added a `Terminate` message to Binary Agreement; otherwise the algorithm would have to run (or at least stay in memory) forever: https://github.com/poanetwork/hbbft/issues/55
* We added a `Conf` message to Binary Agreement: https://github.com/poanetwork/hbbft/issues/37. (The Python implementation will probably also do that.)
* We return some additional information from Subset and Honey Badger, namely which node had input which item/transaction.

Later, we'll implement the dynamic variant that can add and remove nodes (that's not in the paper), and maybe change the "coin schedule" (output fixed 0, fixed 1, then flip the coin, if it actually improves efficiency), and others.

(@vkomenda: Can you think of other points in which we differ from the paper?)

Reply:
This is a good idea. I would like to add a little more context to explain the reasoning behind the modifications (a sentence or so for each). Thanks!
 
- Was pairing-based cryptography chosen to increase security, efficiency and/or implementation ease? Or other reasons? 

- What are potential uses for returned data from Subset & HB? 

- Are there any performance or other potential tradeoffs to run DKG without a trusted dealer? Was this choice made to enhance security/mitigate a potential single point of attack?

I may also add a table which shows the paper's naming conventions compared to this implementation for easy reference.

Reply:
> Rust elliptic curve library

Yes, that was the main reason. If we'd want to switch to e.g. the same elliptic curves that the Python implementation uses, we'd first have to create Rust bindings for some cryptography C library.

> What are potential uses for returned data from Subset & HB?

E.g. you could detect malicious behavior: If a node inputs invalid transactions it may be useful to be able to identify and block or disconnect from that node.

> DKG without a trusted dealer?

Requiring a trusted dealer severely limits its usefulness: In a distributed network you usually don't want a single person to know everyone's secret keys. The paper itself even mentions dealerless DKG, it just doesn't go into details because it goes a bit beyond its scope. The Python implementation [has an open issue for it](https://github.com/amiller/HoneyBadgerBFT/issues/5). The Erlang implementation [has dealerless DKG, too](https://github.com/helium/erlang-dkg).